src/foam/nanos/analytics/CandlestickChartView.js

```diff
@@ -0,0 +1,324 @@
+            this.results[obj.key] = true;
+            this.delegate.put(obj, sub);
+          }
+        }
+      ]
+    }
+  ],
+
+  messages: [
+    {
+      name: 'VIEW_HEADER',
+      message: 'Candlestick Charting'
+    },
+    {
+      name: 'REFRESH_REQUESTED',
+      message: 'Refresh requested'
+    }
+  ],
+
+  properties: [
+    {
+      documentation: `The Candlestick DAO to graph.`,
+      class: 'Reference',
+      of: 'foam.nanos.boot.NSpec',
+      label: 'Candlestick DAO',
+      name: 'candlestickDAOKey',
+      memorable: true,
+      targetDAOKey: 'AuthenticatedNSpecDAO',
+      view: function(_, X) {
+        var E = foam.mlang.Expressions.create();
+        return {
+          class: 'foam.u2.view.RichChoiceView',
+          search: true,
+          sections: [
+            {
+              heading: 'DAO',
+              dao: X.AuthenticatedNSpecDAO
+                .where(E.AND(
+                  E.EQ(foam.nanos.boot.NSpec.SERVE, true),
+                  E.OR(
+                    E.CONTAINS_IC(foam.nanos.boot.NSpec.NAME, "candlestick"),
+                    E.IN(foam.nanos.boot.NSpec.NAME, [
+                      'pmDAO',
+                      'om1MinuteDAO',
+                      'om5MinuteDAO',
+                      'omHouryDAO',
+                      'omDailyDAO'
+                    ])
+                  ),
+                  E.NOT(E.EQ(foam.nanos.boot.NSpec.NAME, "candlestickAlarmDAO"))
+                ))
+                .orderBy(foam.nanos.boot.NSpec.NAME)
+            }
+          ]
+        };
+      },
+      postSet: function(oldValue, newValue) {
+        this.rebuild();
+      }
+    },
+    {
+      documentation: 'Key for first Line dataset.',
+      name: 'candlestickKey1',
+      class: 'String',
+      memorable: true,
+      view: function(_, X) {
+        var dao = X.data.slot(function(candlestickDAOKey) {
+          if ( candlestickDAOKey ) {
+            var self = this;
+            var results = {};
+            return this.SinkDAO.create({
+              delegate: ctrl.__subContext__[candlestickDAOKey].
+                orderBy(self.Candlestick.KEY),
+              sink: this.UniqueSink.create()
+            });
+          }
+          return this.ArrayDAO.create();
+        });
+        return foam.u2.view.ChoiceView.create({
+          objToChoice: function(candlestick) {
+            return [candlestick.key, candlestick.key];
+          },
+          dao$: dao,
+          placeholder: '--'
+        });
+      },
+       postSet: function(oldValue, newValue) {
+        this.rebuild();
+      }
+    },
+    {
+      documentation: 'Key for second Line dataset.',
+      name: 'candlestickKey2',
+      class: 'String',
+      memorable: true,
+      view: function(_, X) {
+        var dao = X.data.slot(function(candlestickDAOKey) {
+          if ( candlestickDAOKey ) {
+            var self = this;
+            var results = {};
+            return this.SinkDAO.create({
```

Comment:
Instead of creating a SinkDAO, which seems weird and unexpected, I would just create a UniqueKeyDOA that extends FilteredDAO but puts a getter: on the "predicate" property to return a predicate that only returns true for unique/first keys. Decorating DAO's rather than sinks seems more expected and would avoid the bug you have with returning the wrong sink.

Reply:
How would a UniqueKeyDAO work without a Sink?

Reply:
addressed in predicate rework.